### DISCUSSION TITLE
How to get the content from a block definition?

### DISCUSSION THREAD

**Discussion by ghemann** [mozman/ezdxf#472] [Q&A]
This is more of a feature request or checking how to use the existing functionality.

I'd like to be able to get all the block definitions and their entities.   So for example, If I have 3 separate block definitions say A, B, and C, and there are 4, 0, and 8 references respectively, is there a way that I can get 3 entities, one for each definition?  I'd like to avoid having to query the whole modelspace for all INSERTS, and that also won't help me find definition B which has no references.  But I would like to extract the entity associated with the definition.   Is there a way to do this?

**Comment by mozman** [comment]
Please specify your question, I don't understand when you mean block references and when you mean the content of blocks.

**Reply by ghemann** [reply]
Sorry, yeah, so I want to extract the geometries/entities of each block definition.   For block references, you can do this by iterating through all entities in modelspace, and if its an insert, i can use virtual_entities() to have access to those entities within the reference.

I'd like to do this with the block definitions, not the block references.     I'd ideally like to avoid having to iterate through all entities.  And also, if I have a definition that doesn't have a reference, I'd like to find that as well.   Let me know if that makes sense and apologies if my terms are incorrect

**Comment by mozman** [comment]
All block definitions are stored in the BLOCKS section of the DXF file. The content of the BLOCKS section is accessible by the attribute `blocks` of the `Drawing` class.

There are two kinds of block definitions:
1. Layouts like model- and paper space layouts
2. Block layouts as reusable geometry blocks

Iteration over all blocks looks like this:

```Python
import ezdxf

doc = ezdxf.readfile("your.dxf")
for layout in doc.blocks:
    process_block_content(layout)

```

Use following properties to check the type of a layout:
- `block.is_block_layout`, block definition for block references (INSERT), xrefs and anonymous blocks (DIMENSION, ACAD_TABLE)
- `block.is_any_layout`, any model- or paper space layout
- `block.is_active_paperspace`, the difference between active and non active paper space layout is only important for CAD applications
- `block.is_any_paperspace`
- `block.is_modelspace`


Next step ist to filter the block definitions from layouts and anonymous blocks:
```Python
for layout in doc.blocks:
    if (
        layout.is_block_layout  # is a block definition?
        and not layout.block.is_anonymous  # layout.block is the BLOCK entity of the block definition
        and not layout.block.is_xref
    ):
        process_block_content(layout)

```

Processing the content entities  is the easy part:
```Python
def process_block_content(block_layout):
    for entity in block_layout:
        process_entity(entity)
```

If you change the content of a block definition, you change the appearance of all block references with this definition!

You cannot know where the block is being used, in model- or paper space, or in any other block definition, just by processing the block definition.

So processing the INSERT entities from a model- or paper space layout is maybe the better way:

```Python
msp = doc.modelspace()
unique_block_names = {insert.name for insert in msp.query("INSERT")}
for name in unique_block_names:
    block_layout = doc.blocks.get(name)
    if (
        not layout.block.is_anonymous
        and not layout.block.is_xref
    ):
        process_block_content(block_layout)
```

And don't worry about processing time, iteration over potentially thousands of entities is fast, even in Python - loading and saving time would be the bigger problem.

**Reply by ghemann** [reply]
This is super helpful, thank you so much!!

**Comment by AndreyB67** [comment]
Goodnight! It looks like you were talking about the process_block_content function and accidentally misspelled by writing the "def process_clock_content"
